File: starters/javascript/tron.js
// If you want to use this file, make sure to include it in your
// submission. You may modify it and submit the modified copy, or you
// may discard it and roll your own.

/**
 * Provided code for the JavaScript starter package
 *
 * See the example bots randbot.js and wallbot.js to get started.
 */

var sys = require('sys');

exports.NORTH = 1;
exports.EAST = 2;
exports.SOUTH = 3;
exports.WEST = 4;

exports.FLOOR = ' ';
exports.WALL  = '#';
exports.ME    = '1';
exports.THEM  = '2';

exports.DIRECTIONS = [exports.NORTH, exports.EAST, exports.SOUTH, exports.WEST];

/* Report an error and exit. */

function invalid_input(message) {
    sys.error("Invalid input: " + message);
    process.exit(1);
}

/* Read stdin, and call the supplied callback function
 * for each complete line of input that is read.
 */

function forEachLineOfInput(callback) {
  var buf = "";
  process.stdio.open("ascii");
  process.stdio.addListener("data",
      function (input) {
          var index, line;
          buf = buf + input;
          while ((index = buf.indexOf('\n')) >= 0) {
              line = buf.substring(0, index+1);
              buf = buf.substring(index+1);
              callback(line);
          }
      }
  );
}

/*
 * play the game. Repeatedly read a board from stdin and then
 * call the moveFn callback function with a "board" object.
 */

exports.play = function(moveFn) {
  var lineNumber = 0;
  var width = 0;
  var height = 0;
  var data = "";
  forEachLineOfInput(
      function (line) {
          var dim;
          line = line.replace(/[\r\n]*$/g,''); // Remove end-of-line chars if present
          if (lineNumber === 0) {
              dim = line.split(" ");
              if (dim.length != 2) {
                  invalid_input("expected width, height on first line, got \"" + line + "\"");
              }
              width = parseInt(dim[0],10);
              height = parseInt(dim[1],10);
              if (width <= 0 || isNaN(width) || height <= 0 || isNaN(height)) {
                  invalid_input("expected width, height on first line, got \"" + line + "\"");
              }
          } else {
              if (line.length != width) {
                  invalid_input("malformed board, line wrong length:" + line.length);
              }
              data = data + line;
              if (lineNumber == height) {
                  sys.puts(moveFn(
                  /* The Board object */
                  {
                    /* width of the board */
                    "width": width,
                    /* height of the board */
                    "height": height,
                    /* the raw data of the board as one long string */
                    "data": data,
                    /* Convert a (y, x) coordinate into an index into the data string. */
                    "YXtoCoord": function (y, x) {
                            if (x < 0 || x >= this.width || y < 0 || y >= this.height ) {
                                return -1;
                            }
                            return y * this.width + x;
                        },
                    /* Extract the x value of a coordinate */
                    "coordX": function(coord) {
                            var y = Math.floor(coord / this.width);
                            return coord - y * this.width;
                        },
                    /* Extract the y value of a coordinate */
                    "coordY": function(coord) {
                            return Math.floor(coord / this.width);
                        },
                    /* Look up the given (y, x) coord in the data string. */
                    "atYX": function (y, x) {
                            return this.data.charAt(this.YXToCoord(y, x));
                        },
                    /* Look up the given index in the data string. */
                    "at": function (coord) {
                            if (this.isOutOfBounds(coord)) {
                                return exports.WALL;
                            }
                            return this.data.charAt(coord);
                        },
                    /* Is a given coordinate out of bounds? */
                    "isOutOfBounds": function (coord) {
                            return coord < 0 || coord >= this.data.length;
                        },
                    /* Find the index of a given element. Throws an exception if missing. */
                    "find": function (item) {
                            var index = this.data.indexOf(item);
                            if (index < 0) {
                                throw "Not found";
                            }
                            return index;
                        },
                    /* Return the index of player 1, also known as "me". */
                    "me": function () {
                            return this.find(exports.ME);
                        },
                    /* Return the index of player 2, also known as "them". */
                    "them": function () {
                            return this.find(exports.THEM);
                        },
                    /* Return true if the given index is passable.
                     * out-of-bounds coordinates are not passible.
                     */
                    "passable": function (index) {
                            if (this.isOutOfBounds(index)) {
                                return false;
                            }
                            return this.data[index] == exports.FLOOR;
                        },
                    /* Compute a new index that is in a relative offset from a given origin.
                     * If you don't pass in an origin, it defaults to "me".
                     * If the new index would cross a boundary of the map
                     * the return index is -1.
                     */
                    "rel": function (direction, origin) {
                            if (origin === undefined) {
                                origin = this.me();
                            }
                            if (direction == exports.NORTH) {
                                if (origin < this.width) {
                                    return -1;
                                }
                                return origin - this.width;
                            } else if (direction == exports.SOUTH) {
                                if (origin >= this.data.length - this.width) {
                                    return -1;
                                }
                                return origin + this.width;
                            } else if (direction == exports.EAST) {
                                if (this.coordX(origin) == this.width - 1) {
                                    return -1;
                                }
                                return origin + 1;
                            } else if (direction == exports.WEST) {
                                if (this.coordX(origin) === 0) {
                                    return -1;
                                }
                                return origin - 1;
                            } else {
                                throw "Invalid direction";
                            }
                         },
                     /* Compute all four adjacent locations to a given origin.
                        If the given location is at an edge, some of the adjacent
                        data will become -1.
                      */
                     "adjacent": function (origin) {
                             return [this.rel(exports.NORTH, origin),
                                 this.rel(exports.EAST, origin),
                                 this.rel(exports.SOUTH, origin),
                                 this.rel(exports.WEST, origin)];
                          },
                      /* Return all the possible legal moves from the player's current
                         location. If there are no legal moves, return [tron.NORTH], just
                         to make it easier for simple algorithms.
                       */
                      "moves": function() {
                              var meCoord = this.me();
                              var adjacents = this.adjacent(meCoord);
                              var moves = [];
                              var i;
                              var coord;
                              var dir;
                              for (i = 0; i < adjacents.length; i++) {
                                  dir = exports.DIRECTIONS[i];
                                  coord = adjacents[i];
                                  if (this.passable(coord)) {
                                      moves.push(dir);
                                  }
                              }
                              if (moves.length === 0) {
                                  // it seems we have already lost
                                  return [exports.NORTH];
                              }
                              return moves;
                          }
                  }));
                  // Reset our accumulators.
                  lineNumber = -1;
                  data = "";
              }
          }
          lineNumber += 1;
      });
};

File: src/tron.py
# If you want to use this file, make sure to include it in your
# submission. You may modify it and submit the modified copy, or you
# may discard it and roll your own.

"""
Provided code for the Python starter package

See the example bots randbot.py and wallbot.py to get started.
"""

import sys, os

def invalid_input(message):
    """You do not need to call this function directly."""

    print >>sys.stderr, "Invalid input: %s" % message
    sys.exit(1)

def readline(buf):
    """You do not need to call this function directly."""

    while not '\n' in buf:
        tmp = os.read(0, 1024)
        if not tmp:
            break
        buf += tmp

    if not buf.strip():
        return None, buf
    if not '\n' in buf:
        invalid_input('unexpected EOF after "%s"' % buf)

    index = buf.find('\n')
    line = buf[0:index]
    rest = buf[index + 1:]
    return line, rest

class Board(object):
    """The Tron Board.

    The recommended way to use this class is as follows:

        def which_move(board):
            # figure this part out yourself
            return tron.NORTH

        for board in tron.Board.generate():
            tron.move(which_move(board))

    Feel free to add stuff to this class.
    """

    def __init__(self, width, height, board):
        """You do not need to call this method directly."""

        self.board = board
        self.height = height
        self.width = width
        self._me = None
        self._them = None

    @staticmethod
    def read(buf):
        """You do not need to call this method directly."""

        meta, buf = readline(buf)

        if not meta:
            return None, buf

        dim = meta.split(' ')

        if len(dim) != 2:
            invalid_input("expected dimensions on first line")

        try:
            width, height = int(dim[0]), int(dim[1])
        except ValueError:
            invalid_input("malformed dimensions on first line")

        lines = []

        while len(lines) != height:
            line, buf = readline(buf)
            if not line:
                invalid_input("unexpected EOF reading board")
            lines.append(line)

        board = [line[:width] for line in lines]

        if len(board) != height or any(len(board[y]) != width for y in xrange(height)):
            invalid_input("malformed board")

        return Board(width, height, board), buf

    @staticmethod
    def generate():
        """Generate board objects, once per turn.

        This method returns a generator which you may iterate over.
        Make sure to call tron.move() exactly once for every board
        generated, or your bot will not work.
        """

        buf = ''

        while True:
            board, buf = Board.read(buf)
            if not board:
                break
            yield board

        if buf.strip():
            invalid_input("garbage after last board: %s" % buf)

    def __getitem__(self, coords):
        """Retrieve the object at the specified coordinates.

        Use it like this:

            if board[3, 2] == tron.THEM:
                # oh no, the other player is at (3,2)
                run_away()

        Coordinate System:
            The coordinate (y, x) corresponds to row y, column x.
            The top left is (0, 0) and the bottom right is
            (board.height - 1, board.width - 1). Out-of-range
            coordinates are always considered walls.

        Items on the board:
            tron.FLOOR - an empty square
            tron.WALL  - a wall or trail of a bot
            tron.ME    - your bot
            tron.THEM  - the enemy bot
        """

        y, x = coords
        if not 0 <= x < self.width or not 0 <= y < self.height:
            return WALL
        return self.board[y][x]

    def me(self):
        """Finds your position on the board.

        It is always true that board[board.me()] == tron.ME.
        """

        if not self._me:
            self._me = self.find(ME)
        return self._me

    def them(self):
        """Finds the other player's position on the board.

        It is always true that board[board.them()] == tron.THEM.
        """

        if not self._them:
            self._them = self.find(THEM)
        return self._them

    def find(self, obj):
        """You do not need to call this method directly."""

        for y in xrange(self.height):
            for x in xrange(self.width):
                if self[y, x] == obj:
                    return y, x
        raise KeyError("object '%s' is not in the board" % obj)

    def passable(self, coords):
        """Determine if a position in the board is passable.

        You can only safely move onto passable tiles, and only
        floor tiles are passable.
        """

        return self[coords] == FLOOR

    def rel(self, direction, origin=None):
        """Calculate which tile is in the given direction from origin.

        The default origin is you. Therefore, board.rel(tron.NORTH))
        is the tile north of your current position. Similarly,
        board.rel(tron.SOUTH, board.them()) is the tile south of
        the other bot's position.
        """

        if not origin:
            origin = self.me()
        y, x = origin
        if direction == NORTH:
            return y - 1, x
        elif direction == SOUTH:
            return y + 1, x
        elif direction == EAST:
            return y, x + 1
        elif direction == WEST:
            return y, x - 1
        else:
            raise KeyError("not a valid direction: %s" % direction)

    def adjacent(self, origin):
        """Calculate the four tiles that are adjacent to origin.

        Particularly, board.adjacent(board.me()) returns the four
        tiles to which you can move to this turn. This does not
        return tiles diagonally adjacent to origin.
        """

        return [self.rel(dir, origin) for dir in DIRECTIONS]

    def moves(self):
        """Calculate which moves are safe to make this turn.

        Any move in the returned list is a valid move. There
        are two ways moving to one of these tiles could end
        the game:

            1. At the beginning of the following turn,
               there are no valid moves off this tile.
            2. The other player also moves onto this tile,
               and you collide.
        """
        possible = dict((dir, self.rel(dir)) for dir in DIRECTIONS)
        passable = [dir for dir in possible if self.passable(possible[dir])]
        if not passable:
            # it seems we have already lost
            return [NORTH]
        return passable

NORTH = 1
EAST  = 2
SOUTH = 3
WEST  = 4

FLOOR = ' '
WALL  = '#'
ME    = '1'
THEM  = '2'

DIRECTIONS = (NORTH, EAST, SOUTH, WEST)

def move(direction):
    print  direction
    sys.stdout.flush()
